infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/StateTransferManager.java

```diff
@@ -0,0 +1,316 @@
+                            .endAddress(segment.getEndAddress())
+                            .status(newStatus)
+                            .build();
+                }
+
+
+        ).collect(ImmutableList.toImmutableList());
+    }
+
+    /**
+     * Given a batch processor and a stream of batch requests, execute the state transfer.
+     *
+     * @param batchProcessor An instance of a batch processor.
+     * @param batchStream    A stream of batch requests.
+     * @return A completed future containing a result of total number of addresses transferred
+     * or an exception.
+     */
+    CompletableFuture<Result<Long, TransferSegmentFailure>> synchronousStateTransfer(
+            StateTransferBatchProcessor batchProcessor, Stream<TransferBatchRequest> batchStream) {
+        Result<Long, TransferSegmentFailure> accumulatedResult = Result.ok(0L);
+
+        Result<Long, TransferSegmentFailure> resultOfTransfer =
+                batchStream.reduce(accumulatedResult, (resultSoFar, nextBatch) -> {
+                    TransferBatchResponse transferBatchResponse =
+                            batchProcessor.transfer(nextBatch).join();
+                    if (transferBatchResponse.getStatus() == FailureStatus.FAILED) {
+                        return Result.error(new TransferSegmentFailure());
+                    } else {
+                        return resultSoFar
+                                .map(sumSoFar -> sumSoFar + transferBatchResponse
+                                        .getTransferBatchRequest()
+                                        .getAddresses()
+                                        .size());
+                    }
+                }, (oldSum, newSum) -> newSum);
+
+        return CompletableFuture.completedFuture(resultOfTransfer);
+    }
+
+
+    /**
+     * Based on a result of a state transfer, update a segment status.
+     *
+     * @param result      A result, containing a total number of records transferred or a failure.
+     * @param totalNeeded A total number of records that must be transferred.
+     * @return An updated segment status.
+     */
+    TransferSegmentStatus createStatusBasedOnTransferResult(
+            Result<Long, TransferSegmentFailure> result, long totalNeeded) {
+
+        Result<Long, TransferSegmentFailure> checkedResult = result.flatMap(totalTransferred -> {
+            if (totalTransferred != totalNeeded) {
+                String error = "Needed: " + totalNeeded + ", but transferred: " + totalTransferred;
+
+                return Result.error(new TransferSegmentFailure(new IllegalStateException(error)));
+            } else {
+                return Result.ok(totalTransferred);
+            }
+        });
+
+        if (checkedResult.isError()) {
+            return TransferSegmentStatus
+                    .builder()
+                    .totalTransferred(0L)
+                    .segmentState(SegmentState.FAILED)
```

Comment:
I think you are not taking care of OverwriteException. If an IO exception happens during local write record, the transfer failed, you set the state to FAILED, and when retrying, the transfer manager won't re-request the known addresses (cuz you only request known addresses if the state is NOT_TRANSFERRED), then you will have OverwriteEception every time you retry. Right?

Reply:
If OWE happens for some reason, what would happen is that the transfer method will immediately fail without the retries in the ProtocolBatchProcessor; 
exceptionally block will wrap this exception in the TransferBatchResponse.
The caller of transfer - a StateTransferManager will fail a synchronous transfer of a segment, the segment will be marked as failed, and the transfer will fail, all state will nullify. On a workflow retry, a segment will be marked as NOT_TRANSFERRED again, the addresses will be re-requested and the exception won't happen again.
However, I agree that there should be some tolerance to OWEs if they do indeed happen as we do now. For example, we can tolerate up to 3 OWEs. I will include that in the impl.